File: go/sigil/content_capture.go
package sigil

import (
	"context"
	"fmt"
	"strings"
)

// ContentCaptureMode controls what content is included in exported generation
// payloads and OTel span attributes.
type ContentCaptureMode int

const (
	// ContentCaptureModeDefault uses the parent or client-level default.
	// On Config this resolves to NoToolContent for backward compatibility.
	// On GenerationStart this inherits from Config.
	// On ToolExecutionStart this inherits from the parent generation context,
	// falling back to Config.
	ContentCaptureModeDefault ContentCaptureMode = iota
	// ContentCaptureModeFull exports all content.
	ContentCaptureModeFull
	// ContentCaptureModeNoToolContent exports full generation content but
	// excludes tool execution content (arguments and results) from span
	// attributes unless explicitly opted in via IncludeContent or a per-tool
	// ContentCapture override. This matches the pre-ContentCaptureMode SDK
	// default behavior.
	ContentCaptureModeNoToolContent
	// ContentCaptureModeMetadataOnly preserves message structure, tool names,
	// usage, and timing but strips text, tool arguments, tool results,
	// thinking, system prompts, conversation titles, and raw artifacts.
	//
	// Note: user-provided Metadata and Tags are NOT stripped — callers are
	// responsible for ensuring these maps do not contain sensitive content
	// when using MetadataOnly mode.
	ContentCaptureModeMetadataOnly
)

const (
	metadataKeyContentCaptureMode        = "sigil.sdk.content_capture_mode"
	contentCaptureModeValueFull          = "full"
	contentCaptureModeValueNoToolContent = "no_tool_content"
	contentCaptureModeValueMetaOnly      = "metadata_only"
)

// resolveContentCaptureMode returns the effective mode from an override and a
// fallback. Default is transparent — it falls through to the fallback.
func resolveContentCaptureMode(override, fallback ContentCaptureMode) ContentCaptureMode {
	if override != ContentCaptureModeDefault {
		return override
	}
	return fallback
}

// callContentCaptureResolver invokes the resolver callback safely, recovering
// from panics. Returns ContentCaptureModeDefault when the resolver is nil.
// Panics are treated as ContentCaptureModeMetadataOnly (fail-closed).
func callContentCaptureResolver(resolver func(ctx context.Context, metadata map[string]any) ContentCaptureMode, ctx context.Context, metadata map[string]any) (mode ContentCaptureMode) {
	if resolver == nil {
		return ContentCaptureModeDefault
	}
	defer func() {
		if r := recover(); r != nil {
			mode = ContentCaptureModeMetadataOnly
		}
	}()
	return resolver(ctx, metadata)
}

// resolveClientContentCaptureMode resolves the effective mode for the client.
// Default at the client level means NoToolContent (backward compatibility):
// generation content is always captured, but tool content requires explicit
// opt-in via IncludeContent or ContentCapture on ToolExecutionStart.
func resolveClientContentCaptureMode(mode ContentCaptureMode) ContentCaptureMode {
	if mode == ContentCaptureModeDefault {
		return ContentCaptureModeNoToolContent
	}
	return mode
}

// stampContentCaptureMetadata sets the content capture mode marker on the generation.
func stampContentCaptureMetadata(g *Generation, mode ContentCaptureMode) {
	if g.Metadata == nil {
		g.Metadata = map[string]any{}
	}
	g.Metadata[metadataKeyContentCaptureMode] = mode.String()
}

// isContentStripped reports whether the generation has been through MetadataOnly
// stripping, based on the stamped metadata marker.
func isContentStripped(g Generation) bool {
	if g.Metadata == nil {
		return false
	}
	v, _ := g.Metadata[metadataKeyContentCaptureMode].(string)
	return v == contentCaptureModeValueMetaOnly
}

// stripContent removes sensitive content from a generation while preserving
// message structure (roles, part kinds), tool names/IDs, usage, timing, and
// all other metadata fields. errorCategory is the classified error category
// (e.g. "rate_limit", "timeout") used to replace the raw CallError text.
func stripContent(g *Generation, errorCategory string) {
	g.SystemPrompt = ""
	g.Artifacts = nil

	if g.CallError != "" {
		if errorCategory != "" {
			g.CallError = errorCategory
		} else {
			g.CallError = "sdk_error"
		}
	}
	delete(g.Metadata, "call_error")

	g.ConversationTitle = ""
	delete(g.Metadata, spanAttrConversationTitle)

	for i := range g.Input {
		stripMessageContent(&g.Input[i])
	}
	for i := range g.Output {
		stripMessageContent(&g.Output[i])
	}
	for i := range g.Tools {
		g.Tools[i].Description = ""
		g.Tools[i].InputSchema = nil
	}
}

func stripMessageContent(m *Message) {
	for i := range m.Parts {
		m.Parts[i].Text = ""
		m.Parts[i].Thinking = ""
		if m.Parts[i].ToolCall != nil {
			m.Parts[i].ToolCall.InputJSON = nil
		}
		if m.Parts[i].ToolResult != nil {
			m.Parts[i].ToolResult.Content = ""
			m.Parts[i].ToolResult.ContentJSON = nil
		}
	}
}

// resolveToolContentCaptureMode resolves the effective content capture mode for
// a tool execution from the per-tool override, parent generation context, and
// client default.
func resolveToolContentCaptureMode(toolMode, ctxMode ContentCaptureMode, ctxSet bool, clientDefault ContentCaptureMode) ContentCaptureMode {
	resolved := resolveClientContentCaptureMode(clientDefault)
	if ctxSet {
		resolved = ctxMode
	}
	if toolMode != ContentCaptureModeDefault {
		resolved = toolMode
	}
	return resolved
}

// String returns the string representation of a ContentCaptureMode.
func (m ContentCaptureMode) String() string {
	switch m {
	case ContentCaptureModeFull:
		return contentCaptureModeValueFull
	case ContentCaptureModeNoToolContent:
		return contentCaptureModeValueNoToolContent
	case ContentCaptureModeMetadataOnly:
		return contentCaptureModeValueMetaOnly
	default:
		return "default"
	}
}

// MarshalText implements encoding.TextMarshaler for ContentCaptureMode.
func (m ContentCaptureMode) MarshalText() ([]byte, error) {
	return []byte(m.String()), nil
}

// UnmarshalText implements encoding.TextUnmarshaler for ContentCaptureMode.
func (m *ContentCaptureMode) UnmarshalText(text []byte) error {
	switch strings.ToLower(string(text)) {
	case contentCaptureModeValueFull:
		*m = ContentCaptureModeFull
	case contentCaptureModeValueNoToolContent:
		*m = ContentCaptureModeNoToolContent
	case contentCaptureModeValueMetaOnly:
		*m = ContentCaptureModeMetadataOnly
	case "default", "":
		*m = ContentCaptureModeDefault
	default:
		return fmt.Errorf("unknown content capture mode: %q", string(text))
	}
	return nil
}

File: js/src/content_capture.ts
import type { ContentCaptureMode, ContentCaptureResolver, Generation, Message } from './types.js';

export const metadataKeyContentCaptureMode = 'sigil.sdk.content_capture_mode';

/**
 * Returns the effective mode from an override and a fallback. `'default'` is
 * transparent — it falls through to the fallback.
 */
export function resolveContentCaptureMode(
  override: ContentCaptureMode,
  fallback: ContentCaptureMode,
): ContentCaptureMode {
  if (override !== 'default') {
    return override;
  }
  return fallback;
}

/**
 * Resolves the effective mode at the client level. `'default'` resolves to
 * `'no_tool_content'` for backward compatibility.
 */
export function resolveClientContentCaptureMode(mode: ContentCaptureMode): ContentCaptureMode {
  if (mode === 'default') {
    return 'no_tool_content';
  }
  return mode;
}

/**
 * Invokes the resolver callback safely, catching thrown errors. Returns
 * `'default'` when the resolver is undefined. Errors are treated as
 * `'metadata_only'` (fail-closed).
 */
export function callContentCaptureResolver(
  resolver: ContentCaptureResolver | undefined,
  metadata: Record<string, unknown> | undefined,
): ContentCaptureMode {
  if (resolver === undefined) {
    return 'default';
  }
  try {
    return resolver(metadata);
  } catch {
    return 'metadata_only';
  }
}

/** Sets the content capture mode marker on the generation metadata. */
export function stampContentCaptureMetadata(generation: Generation, mode: ContentCaptureMode): void {
  if (generation.metadata === undefined) {
    generation.metadata = {};
  }
  generation.metadata[metadataKeyContentCaptureMode] = mode;
}

/**
 * Strips sensitive content from a generation while preserving message
 * structure (roles, part types), tool names/IDs, usage, timing, and all
 * other metadata fields. `errorCategory` is the classified error category
 * used to replace the raw `callError` text.
 */
export function stripContent(generation: Generation, errorCategory: string): void {
  generation.systemPrompt = '';
  generation.conversationTitle = '';
  generation.artifacts = null;

  if (generation.callError !== undefined) {
    generation.callError = errorCategory.length > 0 ? errorCategory : 'sdk_error';
  }
  if (generation.metadata !== undefined) {
    delete generation.metadata.call_error;
    delete generation.metadata['sigil.conversation.title'];
  }

  if (generation.input !== undefined) {
    for (const message of generation.input) {
      stripMessageContent(message);
    }
  }
  if (generation.output !== undefined) {
    for (const message of generation.output) {
      stripMessageContent(message);
    }
  }
  if (generation.tools !== undefined) {
    for (const tool of generation.tools) {
      tool.description = '';
      tool.inputSchemaJSON = '';
    }
  }
}

function stripMessageContent(message: Message): void {
  if (message.parts === undefined) {
    message.content = '';
    return;
  }
  message.content = '';
  for (const part of message.parts) {
    switch (part.type) {
      case 'text':
        part.text = '';
        break;
      case 'thinking':
        part.thinking = '';
        break;
      case 'tool_call':
        part.toolCall.inputJSON = '';
        break;
      case 'tool_result':
        part.toolResult.content = '';
        part.toolResult.contentJSON = '';
        break;
    }
  }
}

/**
 * Determines whether tool execution content (arguments, results) should be
 * included in span attributes. Resolves the effective mode from the
 * explicit tool override, client default, resolver, and legacy
 * `includeContent`.
 */
export function shouldIncludeToolContent(
  toolMode: ContentCaptureMode,
  clientDefault: ContentCaptureMode,
  resolverMode: ContentCaptureMode,
  legacyInclude: boolean,
): boolean {
  let resolved = resolveClientContentCaptureMode(resolveContentCaptureMode(resolverMode, clientDefault));
  if (toolMode !== 'default') {
    resolved = toolMode;
  }
  switch (resolved) {
    case 'metadata_only':
      return false;
    case 'full':
      return true;
    default:
      return legacyInclude;
  }
}
